### ISSUE TITLE
libdhash: hash table subject algorithmic complexity attacks

### ISSUE THREAD

**Issue by pbrezina** [pbrezina/sssd-test#1702]
Cloned from Pagure issue: https://pagure.io/SSSD/sssd/issue/1698
- Created at 2012-12-06 10:45:34 by [jhrozek](https://pagure.io/user/jhrozek)
- Closed at 2017-03-01 17:39:12 as None
- Assigned to nobody
---

https://bugzilla.redhat.com/show_bug.cgi?id=882334 (*Red Hat Enterprise Linux 6*)

    The internal hash function for strings uses this construct:
    
            for (h = 0, k = (unsigned char *) key->str; *k; k++)
                h = h * PRIME_1 ^ (*k - ' ');
    
    It might be a good idea to switch to a keyed hash function (like Bob Jenkins'
    lookup3 function).
    
    I don't know if libdhash is used in ways that expose this and allow attackers
    to mount algorithmic complexity denial-of-service attacks (a few thousand
    entries with attacker-controlled keys would be needed), or if client code
    depends on the predictable iteration order.

### Comments

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2012-12-06 15:36:34](https://pagure.io/SSSD/sssd/issue/1698#416036)**

Fields changed

blockedby:  => 
blocking:  => 
coverity:  => 
design:  => 
design_review:  => 0
feature_milestone:  => 
fedora_test_page:  => 
milestone: NEEDS_TRIAGE => SSSD 1.10 beta
testsupdated:  => 0

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2012-12-06 15:36:34](https://pagure.io/SSSD/sssd/issue/1698#228908)**

Fields changed

blockedby:  => 
blocking:  => 
coverity:  => 
design:  => 
design_review:  => 0
feature_milestone:  => 
fedora_test_page:  => 
milestone: NEEDS_TRIAGE => SSSD 1.10 beta
testsupdated:  => 0

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2012-12-06 15:36:34](https://pagure.io/SSSD/sssd/issue/1698#416466)**

Fields changed

blockedby:  => 
blocking:  => 
coverity:  => 
design:  => 
design_review:  => 0
feature_milestone:  => 
fedora_test_page:  => 
milestone: NEEDS_TRIAGE => SSSD 1.10 beta
testsupdated:  => 0

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2012-12-20 23:33:42](https://pagure.io/SSSD/sssd/issue/1698#416037)**

Fields changed

selected:  => Not need

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2012-12-20 23:33:42](https://pagure.io/SSSD/sssd/issue/1698#416467)**

Fields changed

selected:  => Not need

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2012-12-20 23:33:42](https://pagure.io/SSSD/sssd/issue/1698#228909)**

Fields changed

selected:  => Not need

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2013-01-02 15:32:08](https://pagure.io/SSSD/sssd/issue/1698#228910)**

Moving tickets that are not a priority for SSSD 1.10 into the next release.

milestone: SSSD 1.10 beta => SSSD 1.11 beta

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2013-01-02 15:32:08](https://pagure.io/SSSD/sssd/issue/1698#416468)**

Moving tickets that are not a priority for SSSD 1.10 into the next release.

milestone: SSSD 1.10 beta => SSSD 1.11 beta

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2013-01-02 15:32:08](https://pagure.io/SSSD/sssd/issue/1698#416038)**

Moving tickets that are not a priority for SSSD 1.10 into the next release.

milestone: SSSD 1.10 beta => SSSD 1.11 beta

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2013-01-24 17:06:01](https://pagure.io/SSSD/sssd/issue/1698#416039)**

Just FYI, the current recommended hash function is Siphash, so we will need to implement it before we can address this libdhash issue if we determine we need any change at all.

_comment0: Just FYI, the current recommended hash function is SipHash, so we will need to implement it before we can address this libdhash issue if we determine we need any change at all. => 1359043587232935

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2013-01-24 17:06:01](https://pagure.io/SSSD/sssd/issue/1698#416469)**

Just FYI, the current recommended hash function is Siphash, so we will need to implement it before we can address this libdhash issue if we determine we need any change at all.

_comment0: Just FYI, the current recommended hash function is SipHash, so we will need to implement it before we can address this libdhash issue if we determine we need any change at all. => 1359043587232935

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2013-01-24 17:06:01](https://pagure.io/SSSD/sssd/issue/1698#228911)**

Just FYI, the current recommended hash function is Siphash, so we will need to implement it before we can address this libdhash issue if we determine we need any change at all.

_comment0: Just FYI, the current recommended hash function is SipHash, so we will need to implement it before we can address this libdhash issue if we determine we need any change at all. => 1359043587232935

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2013-07-30 10:53:49](https://pagure.io/SSSD/sssd/issue/1698#228912)**

Fields changed

changelog:  => 
milestone: SSSD 1.13 beta => Tools Deferred
review:  => 0

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2013-07-30 10:53:49](https://pagure.io/SSSD/sssd/issue/1698#416040)**

Fields changed

changelog:  => 
milestone: SSSD 1.13 beta => Tools Deferred
review:  => 0

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2013-07-30 10:53:49](https://pagure.io/SSSD/sssd/issue/1698#416470)**

Fields changed

changelog:  => 
milestone: SSSD 1.13 beta => Tools Deferred
review:  => 0

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2017-02-24 14:39:02](https://pagure.io/SSSD/sssd/issue/1698#228913)**

**Metadata Update from @jhrozek**:
- Issue set to the milestone: Tools Deferred

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2017-03-01 17:38:51](https://pagure.io/SSSD/sssd/issue/1698#418233)**

Ticket is moved out of sssd issue tracker.

https://pagure.io/SSSD/ding-libs/issue/1698

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2017-03-01 17:39:00](https://pagure.io/SSSD/sssd/issue/1698#418234)**

**Metadata Update from @lslebodn**:
- Custom field component reset
- Custom field design_review reset
- Custom field patch reset
- Custom field review reset
- Custom field selected reset
- Custom field testsupdated reset
- Custom field type reset
- Issue close_status updated to: None

---

**Comment from [jhrozek](https://pagure.io/user/jhrozek) at [2017-03-01 17:39:16](https://pagure.io/SSSD/sssd/issue/1698#418235)**

**Metadata Update from @lslebodn**:
- Custom field design_review reset
- Custom field patch reset
- Custom field review reset
- Custom field testsupdated reset
- Issue status updated to: Closed (was: Open)
